### PR TITLE
[BH-1815] Optimalize loading music files in relaxation

### PR DESCRIPTION
**Pull request by lukasmas**
Relaxation is loading smooth with any number of files.

**Your checklist for this pull request**

Make sure that this PR:
- [x] Complies with our guidelines for contributions
- [ ] Has unit tests if possible
- [ ] Has documentation updated
- [x] Has changelog entry added

### CODE REVIEW

File: products/BellHybrid/apps/application-bell-relaxation/model/RelaxationSongsModel.cpp [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+    {
+        return songsRepository->getRecordsCount();
+    }
+    unsigned int RelaxationSongsModel::getMinimalItemSpaceRequired() const
+    {
+        return style::bell_options::h + 2 * style::bell_options::option_margin;
+    }
+    gui::ListItem *RelaxationSongsModel::getItem(gui::Order order)
+    {
+        std::shared_ptr<db::multimedia_files::MultimediaFilesRecord> const sound = getRecord(order);
```
Comment (dawidwojtas-mudita) [inline products/BellHybrid/apps/application-bell-relaxation/model/RelaxationSongsModel.cpp:35]: Use `const auto` instead

File: products/BellHybrid/apps/application-bell-relaxation/model/RelaxationSongsModel.cpp [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+                             activateCallback(*sound);
+                             return true;
+                         },
+                         [=](gui::Item &item) { return true; },
+                         nullptr))
+                        ->build();
+
+        return item;
+    }
+    void RelaxationSongsModel::requestRecords(uint32_t offset, uint32_t limit)
```
Comment (dawidwojtas-mudita) [inline products/BellHybrid/apps/application-bell-relaxation/model/RelaxationSongsModel.cpp:52]: Use `std::uint32_t` here

File: products/BellHybrid/apps/application-bell-relaxation/model/RelaxationSongsModel.hpp [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+        explicit RelaxationSongsModel(ApplicationCommon *application,
+                                      std::unique_ptr<RelaxationSongsRepository> soundsRepository);
+
+        unsigned int requestRecordsCount() override;
+
+        [[nodiscard]] unsigned int getMinimalItemSpaceRequired() const override;
+
+        gui::ListItem *getItem(gui::Order order) override;
+
+        void requestRecords(uint32_t offset, uint32_t limit) override;
```
Comment (dawidwojtas-mudita) [inline products/BellHybrid/apps/application-bell-relaxation/model/RelaxationSongsModel.hpp:45]: `std::uint32_t` ;)

File: products/BellHybrid/apps/application-bell-relaxation/model/RelaxationSongsRepository.cpp [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+    RelaxationSongsRepository::RelaxationSongsRepository(ApplicationCommon *application,
+                                                         std::unique_ptr<app::music::AbstractTagsFetcher> tagsFetcher,
+                                                         const std::map<MusicType, std::string> &pathPrefixes)
+        : app::AsyncCallbackReceiver{application}, application{application},
+          tagsFetcher(std::move(tagsFetcher)), pathPrefixes{pathPrefixes}
+    {}
+
+    void RelaxationSongsRepository::init()
+    {
+        for (const auto &p : pathPrefixes) {
```
Comment (dawidwojtas-mudita) [inline products/BellHybrid/apps/application-bell-relaxation/model/RelaxationSongsRepository.cpp:39]: You can use `const auto &[musicType, musicPath] : pathPrefixies`. It should be more readable in the next line `updateFilesCount(musicType, musicPath);`

**Review by dawidwojtas-mudita — APPROVED** [review]

File: products/BellHybrid/apps/application-bell-relaxation/model/RelaxationSongsModel.cpp [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+        return style::bell_options::h + 2 * style::bell_options::option_margin;
+    }
+    gui::ListItem *RelaxationSongsModel::getItem(gui::Order order)
+    {
+        std::shared_ptr<db::multimedia_files::MultimediaFilesRecord> const sound = getRecord(order);
+        if (!sound) {
+            return nullptr;
+        }
+
+        auto item = (new gui::option::OptionBellMenu(
```
Comment (Lefucjusz) [inline products/BellHybrid/apps/application-bell-relaxation/model/RelaxationSongsModel.cpp:40]: Wouldn't it be cleaner to write it like that?

```
auto item = new gui::option::OptionBellMenu...
return item->build();
````

File: products/BellHybrid/apps/application-bell-relaxation/model/RelaxationSongsModel.hpp [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+    {
+      private:
+        ApplicationCommon *application;
+        std::unique_ptr<RelaxationSongsRepository> songsRepository;
+        OnActivateCallback activateCallback{nullptr};
+
+        bool onMusicListRetrieved(const std::vector<db::multimedia_files::MultimediaFilesRecord> &records,
+                                  unsigned int repoRecordsCount);
+        [[nodiscard]] bool updateRecords(std::vector<db::multimedia_files::MultimediaFilesRecord> records) override;
+        //        int currentOffset = 0;
```
Comment (Lefucjusz) [inline products/BellHybrid/apps/application-bell-relaxation/model/RelaxationSongsModel.hpp:33]: Remove it completely if it's not required anymore :wink:

File: products/BellHybrid/apps/application-bell-relaxation/model/RelaxationSongsRepository.cpp [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+// For licensing, see https://github.com/mudita/MuditaOS/LICENSE.md
+#include "RelaxationSongsRepository.hpp"
+#include <module-db/queries/multimedia_files/QueryMultimediaFilesGetLimited.hpp>
+#include <module-db/queries/multimedia_files/QueryMultimediaFilesCount.hpp>
+#include <AsyncTask.hpp>
+#include <vector>
+
+namespace
+{
+    std::uint32_t calculateOffset(const std::uint32_t offset, const std::uint32_t filesCount)
```
Comment (Lefucjusz) [inline products/BellHybrid/apps/application-bell-relaxation/model/RelaxationSongsRepository.cpp:11]: These consts are not really needed here

File: products/BellHybrid/apps/application-bell-relaxation/model/RelaxationSongsRepository.cpp [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+    {
+        if (offset >= filesCount) {
+            return offset - filesCount;
+        }
+        return offset;
+    }
+    std::pair<std::uint32_t, std::uint32_t> calculateNewOffsetAndLimit(const std::uint32_t offset,
```
Comment (Lefucjusz) [inline products/BellHybrid/apps/application-bell-relaxation/model/RelaxationSongsRepository.cpp:18]: Here neither

File: products/BellHybrid/apps/application-bell-relaxation/model/RelaxationSongsRepository.cpp (hunk 4 above)
Comment (lukasmas) [reply products/BellHybrid/apps/application-bell-relaxation/model/RelaxationSongsRepository.cpp:39]: Good point, I missed this :D

File: products/BellHybrid/apps/application-bell-relaxation/model/RelaxationSongsRepository.cpp [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+        auto task = app::AsyncQuery::createFromQuery(std::move(query), db::Interface::Name::MultimediaFiles);
+        task->setCallback(taskCallback);
+        task->execute(application, this);
+    }
+
+    std::uint32_t RelaxationSongsRepository::getRecordsCount()
+    {
+        std::uint32_t sum{0};
+        std::for_each(
+            filesPerType.begin(), filesPerType.end(), [&sum](const auto &record) mutable { sum += record.second; });
```
Comment (Lefucjusz) [inline products/BellHybrid/apps/application-bell-relaxation/model/RelaxationSongsRepository.cpp:97]: `mutable` seems to be not needed here, also you can write this method as a oneliner using `std::accumulate`:

```
return std::accumulate(filesPerType.begin(), filesPerType.end(), 0, [](const auto sum, const auto &item) { return sum + item.second; });
```

File: products/BellHybrid/apps/application-bell-relaxation/model/RelaxationSongsRepository.cpp [hunk 10]
```diff
@@ -0,0 +1,135 @@
...
+        return sum;
+    }
+
+    void RelaxationSongsRepository::getMusicFiles(
+        std::uint32_t offset,
+        std::uint32_t limit,
+        const RelaxationSongsRepository::OnGetMusicFilesListCallback &callback)
+    {
+        const auto skipViewUpdate = [](const std::vector<db::multimedia_files::MultimediaFilesRecord> &, unsigned int) {
```
Comment (Lefucjusz) [inline products/BellHybrid/apps/application-bell-relaxation/model/RelaxationSongsRepository.cpp:106]: IMO it's better to name all parameters in functions and just explicitly mark them as `[[maybe_unused]]`, you can shorten the arguments declaration by using generic lambda `[](const auto &records, const auto count)`

File: products/BellHybrid/apps/application-bell-relaxation/model/RelaxationSongsRepository.cpp [hunk 11]
```diff
@@ -0,0 +1,135 @@
...
+        std::uint32_t totalFilesCount{0};
+        const auto lastFileIndex = offset + limit;
+
+        if (filesPerType.empty()) {
+            getMusicFiles(MusicType::Relaxation, offset, limit, callback);
+        }
+
+        for (const auto &[type, filesCount] : filesPerType) {
+            totalFilesCount += filesCount;
+            if (auto newOffset = calculateOffset(offset, filesCount); newOffset != offset) {
```
Comment (Lefucjusz) [inline products/BellHybrid/apps/application-bell-relaxation/model/RelaxationSongsRepository.cpp:119]: const

File: products/BellHybrid/apps/application-bell-relaxation/model/RelaxationSongsRepository.cpp [hunk 12]
```diff
@@ -0,0 +1,135 @@
...
+                offset = newOffset;
+            }
+            else if (lastFileIndex <= totalFilesCount) {
+                getMusicFiles(type, offset, limit, callback);
+                break;
+            }
+            else {
+                auto filesToLoad = filesCount - offset;
```
Comment (Lefucjusz) [inline products/BellHybrid/apps/application-bell-relaxation/model/RelaxationSongsRepository.cpp:127]: const

File: products/BellHybrid/apps/application-bell-relaxation/model/RelaxationSongsRepository.hpp [hunk 13]
```diff
@@ -0,0 +1,42 @@
+// Copyright (c) 2017-2023, Mudita Sp. z.o.o. All rights reserved.
+// For licensing, see https://github.com/mudita/MuditaOS/LICENSE.md
+
+#pragma once
+#include <apps-common/models/SongsRepository.hpp>
```
Comment (Lefucjusz) [inline products/BellHybrid/apps/application-bell-relaxation/model/RelaxationSongsRepository.hpp:5]: Please separate `#pragma`s from `#include`s and namespace declarations by an empty line, that way it's easier to read

**Review by Lefucjusz — APPROVED** [review]

File: products/BellHybrid/apps/application-bell-relaxation/model/RelaxationSongsRepository.cpp (hunk 7 above)
Comment (lukasmas) [reply products/BellHybrid/apps/application-bell-relaxation/model/RelaxationSongsRepository.cpp:11]: We should discuss it with a team :D 
I think this is a good practice to keep this a constant. It protects you from unwanted updates of the value.
We can also consider passing as a const& even for ints. The cost and time are the same :D

File: products/BellHybrid/apps/application-bell-relaxation/model/RelaxationSongsRepository.cpp (hunk 9 above)
Comment (lukasmas) [reply products/BellHybrid/apps/application-bell-relaxation/model/RelaxationSongsRepository.cpp:97]: Thanks, I didn't know this function :D

File: products/BellHybrid/apps/application-bell-relaxation/model/RelaxationSongsRepository.cpp (hunk 7 above)
Comment (lukasmas) [reply products/BellHybrid/apps/application-bell-relaxation/model/RelaxationSongsRepository.cpp]: Discussed on slack.

**Pull request merged**